### PR TITLE
Fix R2DBC with OAuth

### LINKED ISSUE #13467: Creating a blog app with reactive + MySQL doesn't compile
##### **Overview of the issue**

If you create a simple blog monolith with reactive and MySQL, it fails to compile.

```
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ blog ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 68 source files to /Users/mraible/dev/choose-mono/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/mraible/dev/choose-mono/src/main/java/com/jhipster/demo/blog/repository/BlogRepositoryInternalImpl.java:[5,51] cannot find symbol
  symbol:   class UserRowMapper
  location: package com.jhipster.demo.blog.repository.rowmapper
[ERROR] /Users/mraible/dev/choose-mono/src/main/java/com/jhipster/demo/blog/repository/BlogRepositoryInternalImpl.java:[40,19] cannot find symbol
  symbol:   class UserRowMapper
  location: class com.jhipster.demo.blog.repository.BlogRepositoryInternalImpl
[ERROR] /Users/mraible/dev/choose-mono/src/main/java/com/jhipster/demo/blog/repository/BlogRepositoryInternalImpl.java:[50,9] cannot find symbol
  symbol:   class UserRowMapper
  location: class com.jhipster.demo.blog.repository.BlogRepositoryInternalImpl
[INFO] 3 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

##### **Motivation for or Use Case**

I like to show off new features in JHipster. ;)

##### **Reproduce the error**

Create an app with the following JDL, then run `./mvnw`.

```
application {
  config {
    baseName blog
    applicationType monolith
    authenticationType oauth2
    packageName com.jhipster.demo.blog
    prodDatabaseType mysql
    cacheProvider hazelcast
    buildTool maven
    clientFramework vue
    reactive true
    testFrameworks [cypress]
  }
  entities *
}

entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}

entity Post {
  title String required
  content TextBlob required
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{entry}
}

paginate Post, Tag with infinite-scroll
```

##### **Suggest a Fix**

It looks like the `UserRowMapper` is missing, so maybe it just needs to be added?

image

##### **JHipster Version(s)**

7.0 beta 0

### PR DESCRIPTION
**Pull request by mraible**
Should fix #13467, eventually.

---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

Status: Jan 12 21:56 MST, I get the following error when I try to view or edit a blog entity.

```
2021-01-12 21:50:30.846 ERROR 86651 --- [ctor-http-nio-3] o.z.problem.spring.common.AdviceTraits   : Internal Server Error

org.springframework.dao.IncorrectResultSizeDataAccessException: Query [SELECT e.id AS e_id, e.name AS e_name, e.handle AS e_handle, e.user_id AS e_user_id, e_user.id AS user_id, e_user.login AS user_login, e_user.first_name AS user_first_name, e_user.last_name AS user_last_name, e_user.email AS user_email, e_user.activated AS user_activated, e_user.lang_key AS user_lang_key, e_user.image_url AS user_image_url FROM blog e LEFT OUTER JOIN jhi_user AS e_user ON e.user_id = e_user.id] returned non unique result.
	at org.springframework.data.r2dbc.core.DefaultFetchSpec.lambda$one$0(DefaultFetchSpec.java:61)
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException:
Assembly trace from producer [reactor.core.publisher.FluxFlatMap] :
	reactor.core.publisher.Flux.flatMap(Flux.java:4927)
	org.springframework.data.r2dbc.core.DefaultFetchSpec.one(DefaultFetchSpec.java:53)
Error has been observed at the following site(s):
	|_       Flux.flatMap ⇢ at org.springframework.data.r2dbc.core.DefaultFetchSpec.one(DefaultFetchSpec.java:53)
	|_          Flux.next ⇢ at org.springframework.data.r2dbc.core.DefaultFetchSpec.one(DefaultFetchSpec.java:65)
	|_ Mono.switchIfEmpty ⇢ at tech.jhipster.web.util.reactive.ResponseUtil.wrapOrNotFound(ResponseUtil.java:36)
	|_           Mono.map ⇢ at tech.jhipster.web.util.reactive.ResponseUtil.wrapOrNotFound(ResponseUtil.java:37)
	|_     Mono.usingWhen ⇢ at org.springframework.transaction.interceptor.TransactionAspectSupport$ReactiveTransactionSupport.lambda$null$3(TransactionAspectSupport.java:870)
```

Here's the JDL I'm testing with:

```
application {
  config {
    baseName blog
    applicationType monolith
    authenticationType oauth2
    packageName com.jhipster.demo.blog
    prodDatabaseType mysql
    buildTool maven
    clientFramework vue
    reactive true
    testFrameworks [cypress]
  }
  entities *
}

entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}

entity Post {
  title String required
  content TextBlob required
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{entry}
}

paginate Post, Tag with infinite-scroll
```

### CODE REVIEW

**Comment by mraible** [comment]
Related: When I tried to view the H2 database to run the query above, I found it returns an error.

Screen Shot 2021-01-12 at 10 14 09 PM

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs [hunk 1]
```diff
@@ -389,7 +389,7 @@ relationships.forEach((relationship, idx) => {
 
     <%_ if (databaseType === 'sql' && reactive) { _%>
     @Column("<%= getColumnName(relationshipName) %>_id")
-    private Long <%= relationshipFieldName %>Id;
+    private <% if (relationshipFieldName === 'user' && authenticationType === 'oauth2') { %>String<% } else { %>Long<% } %> <%= relationshipFieldName %>Id;
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs:392]:
```suggestion
    private <% relationship.otherEntity.primaryKey.type %> <%= relationshipFieldName %>Id;
```

**Comment by pascalgrimaud** [comment]
Sorry @mshima : I merged it too quickly :)

**Comment by mshima** [comment]
@pascalgrimaud don't worry it’s just an improvement to support custom type otherwise it’s correct.

**Comment by atomfrede** [comment]
> Related: When I tried to view the H2 database to run the query above, I found it returns an error.
> 
> Screen Shot 2021-01-12 at 10 14 09 PM

I think this is because the way we start the h2 console are only working in non reactive web environments. For micronaut we had a similar issues (as we are not running on spring) and started the h2 webserver manually instead of the h2 servlet. I guess we can do the same thing here for reactive part. See https://github.com/jhipster/generator-jhipster-micronaut/blob/59931d1c85ec36141622e18a288f68cd11b205c2/generators/server/templates/src/main/java/package/App.java.ejs#L108

**Comment by DanielFran** [comment]
@atomfrede This seems a candidate to add to jhipster-bom like it is today for non reactive web environments

**Pull request merged**